File: driftwatch/backend/internal/api/agent.go
package api

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx/v5"

	"github.com/souvik-biswas-dev/driftwatch/internal/docker"
)

// agentKeyHeader carries the plaintext agent key when an agent pushes live
// Docker state. The backend only ever stores its SHA-256 hash.
const agentKeyHeader = "X-DriftWatch-Agent-Key"

// hashAgentKey returns the lowercase hex SHA-256 of an agent key.
func hashAgentKey(key string) string {
	sum := sha256.Sum256([]byte(key))
	return hex.EncodeToString(sum[:])
}

// generateAgentKey returns a new random agent key (prefixed "dw_").
func generateAgentKey() (string, error) {
	b := make([]byte, 24)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return "dw_" + hex.EncodeToString(b), nil
}

// handleAgentState ingests a live Docker snapshot pushed by a project's agent.
// The agent runs on the user's own host, so the backend never reaches into their
// Docker daemon directly.
func (a *API) handleAgentState(c *gin.Context) {
	key := c.GetHeader(agentKeyHeader)
	if key == "" {
		respondError(c, http.StatusUnauthorized, "missing agent key", "NO_AGENT_KEY")
		return
	}

	project, err := a.queries.GetProjectByAgentKeyHash(c.Request.Context(), hashAgentKey(key))
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			respondError(c, http.StatusUnauthorized, "invalid agent key", "BAD_AGENT_KEY")
			return
		}
		respondError(c, http.StatusInternalServerError, "lookup failed", "DB_ERROR")
		return
	}

	var live docker.LiveSnapshot
	if err := c.ShouldBindJSON(&live); err != nil {
		respondError(c, http.StatusBadRequest, err.Error(), "INVALID_BODY")
		return
	}

	// Run the scan asynchronously so the agent gets a fast acknowledgement.
	go a.scheduler.IngestLiveState(project.ID, &live)

	respond(c, http.StatusAccepted, gin.H{"project_id": project.ID}, "state accepted")
}

File: driftwatch/backend/src/api/routes/agent.ts
import { Router } from 'express';
import { z } from 'zod';

import { getProjectByAgentKeyHash } from '../../db/projects.js';
import { logger } from '../../logger.js';
import type { LiveSnapshot } from '../../types.js';
import { AGENT_KEY_HEADER, hashAgentKey } from '../agentKey.js';
import type { ApiDeps } from '../deps.js';
import { respond, respondError } from '../respond.js';
import { formatZodError } from '../validate.js';

/**
 * Deliberately permissive: an older agent that omits a field should still be
 * accepted rather than 400'd out of monitoring.
 */
const containerStateSchema = z.object({
	name: z.string().default(''),
	image: z.string().default(''),
	env: z.record(z.string(), z.string()).default({}),
	ports: z.array(z.string()).default([]),
	running: z.boolean().default(false)
});

const liveSnapshotSchema = z.object({
	containers: z.array(containerStateSchema).default([]),
	captured_at: z.string().default(() => new Date().toISOString())
});

/**
 * Ingests a live Docker snapshot pushed by a project's agent. The agent runs on
 * the user's own host, so the backend never reaches into their Docker daemon.
 */
export function agentRoutes(deps: ApiDeps): Router {
	const r = Router();

	r.post('/agent/state', async (req, res) => {
		const key = req.get(AGENT_KEY_HEADER);
		if (!key) {
			respondError(res, 401, 'missing agent key', 'NO_AGENT_KEY');
			return;
		}

		let project;
		try {
			project = await getProjectByAgentKeyHash(deps.db, hashAgentKey(key));
		} catch {
			respondError(res, 500, 'lookup failed', 'DB_ERROR');
			return;
		}
		if (!project) {
			respondError(res, 401, 'invalid agent key', 'BAD_AGENT_KEY');
			return;
		}

		const parsed = liveSnapshotSchema.safeParse(req.body);
		if (!parsed.success) {
			respondError(res, 400, formatZodError(parsed.error), 'INVALID_BODY');
			return;
		}
		const live: LiveSnapshot = parsed.data;

		// Run the scan in the background so the agent gets a fast acknowledgement.
		void deps.scheduler.ingestLiveState(project.id, live).catch((err) => {
			logger.error('agent ingest failed', { project_id: project.id, error: err });
		});

		respond(res, 202, { project_id: project.id }, 'state accepted');
	});

	return r;
}
